--- old/ext/recode/recode.c	2008-02-21 06:40:08.000000000 +0100
+++ new/ext/recode/recode.c	2008-02-21 06:40:34.000000000 +0100
@@ -132,11 +132,11 @@
 {
 	RECODE_REQUEST request = NULL;
 	char *r = NULL;
-	int r_len = 0, r_alen = 0;
+	size_t r_len = 0, r_alen = 0;
 	int req_len, str_len;
 	char *req, *str;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "ss", &req, &req_len, &str, &str_len) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "ss", &req, &req_len, &str, &str_len) == FAILURE || str_len < 0) {
 		return;
 	}
 
