commit 6ed16753c146ea2a06271ac537761430bad3059a
Author: Xinchen Hui <laruence@php.net>
Date:   Sat Apr 20 15:23:49 2013 +0800

    Fixed bug #64677 (execution operator `` stealing surrounding arguments)

--- /dev/null
+++ b/Zend/tests/bug64677.phpt
@@ -0,0 +1,21 @@
+--TEST--
+Bug #64677 (execution operator `` stealing surrounding arguments)
+--FILE--
+<?PHP
+class cat {
+	public function show_output($prepend, $output = '') {
+	}
+}
+$cat = new cat();
+$cat->show_output('Files: ', trim(`cd .`)); // this gives invalid args to shell_exec
+$cat->show_output('Files: ', `cd .`); // this causes a segmentation fault
+$cat->show_output(`cd .`); // this causes a segmentation fault
+
+function show_outputa($prepend, $output) {
+	echo "Okey";
+}
+show_outputa('Files: ', `cd .`); // this works as expected
+
+?>
+--EXPECT--
+Okey
--- a/Zend/zend_compile.c
+++ b/Zend/zend_compile.c
@@ -5651,6 +5651,7 @@
 	GET_CACHE_SLOT(opline->op1.constant);
 	opline->extended_value = 1;
 	SET_UNUSED(opline->op2);
+	opline->op2.num = CG(context).nested_calls;
 	GET_NODE(result, opline->result);
 }
 /* }}} */
