Index: php/ext/odbc/php_odbc_includes.h
===================================================================
--- php.orig/ext/odbc/php_odbc_includes.h
+++ php/ext/odbc/php_odbc_includes.h
@@ -230,8 +230,8 @@ typedef struct odbc_connection {
 typedef struct odbc_result_value {
 	char name[32];
 	char *value;
-	SDWORD vallen;
-	SDWORD coltype;
+	SQLLEN vallen;
+	SQLLEN coltype;
 } odbc_result_value;
 
 typedef struct odbc_result {
