--- a/Archive/Tar.php	2012-02-23 11:31:26.907422679 +0000
+++ b/Archive/Tar.php	2012-02-23 11:32:14.171462803 +0000
@@ -1736,7 +1736,7 @@
             if ($this->_compress_type == 'gz') {
                 while (!@gzeof($v_temp_tar)) {
                     $v_buffer = @gzread($v_temp_tar, 512);
-                    if ($v_buffer == ARCHIVE_TAR_END_BLOCK) {
+                    if ($v_buffer == ARCHIVE_TAR_END_BLOCK || strlen($v_buffer) == 0) {
                         // do not copy end blocks, we will re-make them
                         // after appending
                         continue;
