diff --git a/ext/libxml/libxml.c b/ext/libxml/libxml.c
index e42d845..a39c875 100644
--- a/ext/libxml/libxml.c
+++ b/ext/libxml/libxml.c
@@ -677,9 +677,18 @@ is_string:
 static xmlParserInputPtr _php_libxml_pre_ext_ent_loader(const char *URL,
 		const char *ID, xmlParserCtxtPtr context)
 {
+	TSRMLS_FETCH();
+
 	/* Check whether we're running in a PHP context, since the entity loader
-	 * we've defined is an application level (true global) setting */
-	if (xmlGenericError == php_libxml_error_handler) {
+	 * we've defined is an application level (true global) setting.
+	 * If we are, we also want to check whether we've finished activating
+	 * the modules (RINIT phase). Using our external entity loader during a
+	 * RINIT should not be problem per se (though during MINIT it is, because
+	 * we don't even have a resource list by then), but then whether one
+	 * extension would be using the custom external entity loader or not
+	 * could depend on extension loading order
+	 * (if _php_libxml_per_request_initialization */
+	if (xmlGenericError == php_libxml_error_handler && PG(modules_activated)) {
 		return _php_libxml_external_entity_loader(URL, ID, context);
 	} else {
 		return _php_libxml_default_entity_loader(URL, ID, context);
