commit b980e4fedc945dd0a45d73349716f6ab18a97635
Author: Patrick Allaert <pa@ez.no>
Date:   Tue Jul 9 15:05:10 2013 +0200

    Fixed #65225: PHP_BINARY incorrectly set

--- a/main/main.c
+++ b/main/main.c
@@ -115,6 +115,10 @@
 #endif
 /* }}} */
 
+#ifndef S_ISREG
+#define S_ISREG(mode) (((mode) & S_IFMT) == S_IFREG)
+#endif
+
 PHPAPI int (*php_register_internal_extensions_func)(TSRMLS_D) = php_register_internal_extensions;
 
 #ifndef ZTS
@@ -276,13 +280,14 @@
 			if ((envpath = getenv("PATH")) != NULL) {
 				char *search_dir, search_path[MAXPATHLEN];
 				char *last = NULL;
+				struct stat s;
 
 				path = estrdup(envpath);
 				search_dir = php_strtok_r(path, ":", &last);
 
 				while (search_dir) {
 					snprintf(search_path, MAXPATHLEN, "%s/%s", search_dir, sapi_module.executable_location);
-					if (VCWD_REALPATH(search_path, binary_location) && !VCWD_ACCESS(binary_location, X_OK)) {
+					if (VCWD_REALPATH(search_path, binary_location) && !VCWD_ACCESS(binary_location, X_OK) && VCWD_STAT(binary_location, &s) == 0 && S_ISREG(s.st_mode)) {
 						found = 1;
 						break;
 					}
